File: jlogkim/066.java
class Solution {
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;
        
        boolean existance[] = new boolean [26];
        int prefix[] = new int[26];
        
        for(int i=0; i <26; i++){
            prefix[i]=-1;
        }
        
        int pre=skill.charAt(0)-'A';

        int now;
        for(int i=1; i <skill.length(); i++){
            now=skill.charAt(i)-'A';
            prefix[now]=pre;
            pre=now;
        }
        String target;
        boolean flag=true;
        for(int tc=0; tc<skill_trees.length; tc++){
            flag=true;
            for(int i=0;i<26;i++){
                existance[i]=false;
            }
            target = skill_trees[tc];
            for(int i=0;i<target.length();i++){
                if(prefix[target.charAt(i)-'A']>=0&&existance[prefix[target.charAt(i)-'A']]==false){flag=false;
                        break;                                                                           }
                existance[target.charAt(i)-'A']=true;
            }
            if(flag)answer++;                
        }        
        
        return answer;
    }
}

File: joonas/066.cpp
#include <string>
#include <vector>
#include <map>
using namespace std;

int solution(string skill, vector<string> skill_trees) {
    map<char, int> lv;
    for (int i = 0; i < skill.length(); ++i) {
        lv[skill[i]] = i;
    }
    
    int answer = 0;
    for (auto& t : skill_trees) {
        bool ans = true;
        
        int lastLv = -1;
        for (auto& c : t) {
            if (lv.find(c) == lv.end()) continue;
            if (lastLv + 1 != lv[c]) {
                ans = false;
                break;
            }
            lastLv = lv[c];
        }
        
        answer += ans;
    }
    return answer;
}